src/main/java/com/microsoft/azure/relay/HybridConnectionListener.java

```diff
@@ -277,50 +276,46 @@ public HybridConnectionListener(String connectionString, String path) throws URI
 	 * @return A CompletableFuture which completes when all connections are
 	 *         disconnected with the cloud service
 	 */
+	@SuppressWarnings("unchecked")
 	public CompletableFuture<Void> closeAsync(Duration timeout) {
-		try {
-			CompletableFutureUtil.timedRunAsync(timeout, () -> {
-					List<HybridConnectionChannel> clients;
-					synchronized (this.thisLock) {
-						if (this.closeCalled) {
-							return;
-						}
-	
-						RelayLogger.logEvent("closing", this);
-						this.closeCalled = true;
-	
-						// If the input queue is empty this completes all pending waiters with null and
-						// prevents any new items being added to the input queue.
-						this.connectionInputQueue.shutdown();
-	
-						// Close any unaccepted rendezvous. DequeueAsync won't block since we've called
-						// connectionInputQueue.Shutdown().
-						clients = new ArrayList<HybridConnectionChannel>(this.connectionInputQueue.getPendingCount());
-						HybridConnectionChannel webSocket;
-						do {
-							webSocket = this.connectionInputQueue.dequeueAsync().join();
-	
-							if (webSocket != null) {
-								clients.add(webSocket);
-							}
-						} while (webSocket != null);
-	
-						this.isOnline = false;
-					}
-	
-					clients.forEach(client -> client.closeAsync(
-							new CloseReason(CloseReason.CloseCodes.NORMAL_CLOSURE, "Client closing the socket normally")));
-	
-					RelayLogger.logEvent("closed", this);
-				},
-				EXECUTOR).join();
-		}
-		catch (Exception e) {
-			throw RelayLogger.throwingException(e, this);
-		} finally {
+		return CompletableFutureUtil.timedSupplyAsync(timeout, () -> {
+			synchronized (this.thisLock) {
+				if (this.closeCalled) {
+					return CompletableFuture.completedFuture(null);
+				}
+
+				RelayLogger.logEvent("closing", this);
+				this.closeCalled = true;
+
+				// If the input queue is empty this completes all pending waiters with null and
+				// prevents any new items being added to the input queue.
+				this.connectionInputQueue.shutdown();
+
+				// Close any unaccepted rendezvous. DequeueAsync won't block since we've called
+				// connectionInputQueue.Shutdown().
+				CompletableFuture<?>[] closeTasks = new CompletableFuture<?>[this.connectionInputQueue.getPendingCount()];
```

Comment:
>CompletableFuture<?>[] closeTasks = new CompletableFuture<?>[this.connectionInputQueue.getPendingCount()]; [](start = 4, length = 106)

I'm just curious, does List make this any prettier?

Reply:
CompletableFuture.allof() doesn't take List

---
In reply to: [257775841](https://github.com/Azure/azure-relay-java/pull/14#discussion_r257775841) [](ancestors = 257775841)

Reply:
fair enough

---
In reply to: [258613192](https://github.com/Azure/azure-relay-java/pull/14#discussion_r258613192) [](ancestors = 258613192,257775841)